### PR TITLE
Rework the waiting menu and fix 'calendar::print_duration()'

### PR DESCRIPTION
**Pull request by codemime**
So that it:
- Looks better in the code
- Shows the actual time to wait if player has a watch
- Shows the current game time if player has a watch
- ~~Disables certain menu items if the target time is reached, but not yet passed (for 30 minutes).~~

Also, this PR fixes `calendar::print_duration()` that was broken. It was putting bare turns into the "fraction parts" of the time instead of the right time values, so that you would see e.g.: `12 hours and 476 minutes`. Also, it could never show you round units of time (`1 hour`, `1 minute` e.t.c).

**Before:**
before

**After:**
after_2

### CODE REVIEW

**Comment by Night-Pryanik** [comment]
I agree, that's dubious, but what if it's noon, and player wants to wait till noon of the next day?

**Comment by Zireael07** [comment]
@Night-Pryanik: I don't think it's a valid use case and it would require reworking the menu to have the option to wait for the *next* noon.

**Comment by Night-Pryanik** [comment]
@Zireael07 Example: It's 12:00, noon. I want to wait exactly 24 hours, till next noon, but now it's not possible.

**Comment by Zireael07** [comment]
I know what you want. The old menu didn't enable this and I think it's out of scope for this PR.

**Comment by codemime** [comment]
I agree with @Zireael07 that it's hardly the case. Prior to #17106 the waiting (till the noon in particular) was broken (due to an erroneous formula), but nobody complained about it (which proves that people don't like waiting :) . Even properly working, the old wait menu (prior to #17106) uses the same logic, but in a more obscure way: it doesn't grey out the menu item, but stops waiting immediately if the current hour and the desired hour are equal. In case of noon you had to use another option if the time was between 12:00pm and 01:00pm.

Greyed out menu items signal that the desired time is now. I can even add an appropriate note, something like `Wait till dawn (which is now)` to make that more clear. 

Waiting till the next noon doesn't require any significant reworking, just dropping a couple of lines maybe. If the community decides that this limitation is annoying, I could: reduce the time of disabling (say, to 10 minutes), or even drop it entirely, or replace with a pair `"Wait till the forthcoming noon"` / `"Wait till the next noon"`. It's discussable.

> I want to wait exactly 24 hours, till next noon, but now it's not possible.

_Now_ is an incorrect word in the sentence. Have you ever tried that? That wasn't possible in the previous menu either (see above), and you had to wait an hour instead of a half.

File: src/calendar.cpp [hunk 1]
```diff
@@ -1,12 +1,17 @@
...
+
 #include "calendar.h"
 #include "output.h"
 #include "options.h"
 #include "translations.h"
 #include "game.h"
 #include "debug.h"
 
+// Divided by 100 to prevent overflowing when converted to turns
+const calendar calendar::INDEFINITELY_LONG( int( std::numeric_limits<int>::max() / 100 ) );
```
Comment (BevapDin) [inline src/calendar.cpp]: Why the cast to int, shouldn't it already be an int? Also, the comment is not clear. The calendar constructor takes turns as input, there is no conversion there, did you mean moves (as in `Creature::moves`)?

File: src/calendar.cpp [hunk 2]
```diff
@@ -27,6 +32,25 @@ bool calendar::eternal_season = false;
 // How long, in seconds, does sunrise/sunset last?
 #define TWILIGHT_SECONDS (60 * 60)
 
+constexpr int FULL_SECONDS_IN( int n )
```
Comment (BevapDin) [inline src/calendar.cpp]: I don't get the function name. Full seconds in what?

File: src/game.cpp [hunk 3]
```diff
@@ -13883,72 +13883,56 @@ void game::spawn_mon(int /*shiftx*/, int /*shifty*/)
...
-    if (bHasWatch) {
-        as_m.entries.push_back(uimenu_entry(3, true, '3', _("1 hour")));
-        as_m.entries.push_back(uimenu_entry(4, true, '4', _("2 hours")));
-        as_m.entries.push_back(uimenu_entry(5, true, '5', _("3 hours")));
-        as_m.entries.push_back(uimenu_entry(6, true, '6', _("6 hours")));
+    const bool has_watch = u.has_watch();
+    const auto add_menu_item = [ &as_m, &durations, has_watch ]
+        ( int retval, int hotkey, const std::string &caption = "", int duration = calendar::INDEFINITELY_LONG ) {
+
+        const bool enabled = duration == calendar::INDEFINITELY_LONG || duration < DAYS( 1 ) - MINUTES( 30 );
```
Comment (BevapDin) [inline src/game.cpp]: This strikes me as bad UI. An option should only be disabled if it is impossible to execute it (as in "division by zero" or similar). Whether it's useful to chose that option should be left to the player. It is certainly possible to wait (nearly) 24 hours, why not allow it.

You already show the current time and the time to wait for those options (great!), that should be enough to allow the player to decide as they like.

Simply put: why prevent the player from choosing an option that the game can easily handle?

File: src/game.cpp [hunk 4]
```diff
@@ -13883,72 +13883,56 @@ void game::spawn_mon(int /*shiftx*/, int /*shifty*/)
...
+        std::string text( caption );
+
+        if( enabled && has_watch && duration != calendar::INDEFINITELY_LONG ) {
+            const std::string dur_str( calendar::print_duration( duration ) );
+            text += ( text.empty() ? dur_str : string_format( " (%s)", dur_str.c_str() ) );
+        }
+        as_m.entries.push_back( uimenu_entry( retval, enabled, hotkey, text ) );
```
Comment (BevapDin) [inline src/game.cpp]: You can use `addentry` directly instead of creating a temporary `uimenu_entry`.

File: src/calendar.cpp (hunk 1 above)
Comment (codemime) [inline src/calendar.cpp]: Yes to both questions.

File: src/calendar.cpp (hunk 2 above)
Comment (codemime) [inline src/calendar.cpp]: Uh, in the argument?

File: src/game.cpp (hunk 4 above)
Comment (codemime) [inline src/game.cpp]: A piece of pre-existed code. The `uimenu` class encourages this kind of code (I mean `entries` most definitely should be encapsulated, otherwise it's a design flaw). The replacement is reasonable though.

**Comment by Coolthulhu** [comment]
Needs a rebase

**Comment by codemime** [comment]
Rebased against `master`. All valid options are enabled.
